jms/org.wso2.carbon.transport.jms/src/test/java/org/wso2/carbon/transport/jms/test/QueueTopicClientAckListeningTestCase.java

```diff
@@ -0,0 +1,136 @@
+
+        // Create a topic transport listener
+        jmsTopicTransportListener = new JMSServerConnector("2");
+        topicTestMessageProcessor = new TestMessageProcessor();
+        jmsTopicTransportListener.setMessageProcessor(topicTestMessageProcessor);
+    }
+
+    /**
+     * JMS Server connector will start listening to a queue in client ack mode and there will be a publisher publishing
+     * to the queue. For some messages call acknowledge, reject some messages so that session recover can be called.
+     * After publishing the messages to queue, check whether the number of received messages are greater than the
+     * sent messages, since the jms provider need to re-deliver the messages when session recover is called.
+     */
+    @Test(groups = "jmsListening", description = "Testing whether queue listening is working correctly without any "
+            + "exceptions in client ack mode")
+    public void queueListeningTestCase() {
+        try {
+            jmsQueueTransportListener.start(queueListeningParameters);
+            logger.info("JMS Transport Listener is starting to listen to the queue " + JMSTestConstants.QUEUE_NAME_2);
+            jmsServer.publishMessagesToQueue(JMSTestConstants.QUEUE_NAME_2);
+            Assert.assertTrue(queueTestMessageProcessor.getCount() > 10,
+                    "Expected message count is not received when " + "listening to queue "
+                            + JMSTestConstants.QUEUE_NAME_2);
+            jmsQueueTransportListener.stop();
+        } catch (Exception e) {
+            Assert.fail("Error while listening to queue");
+        }
+    }
+
+    /**
+     * JMS Server connector will start listening to a topic in client ack mode and there will be a publisher publishing
+     * to the topic. For some messages call acknowledge, reject some messages so that session recover can be called.
+     * After publishing the messages to topic, check whether the number of received messages are greater than the
+     * sent messages, since the jms provider need to re-deliver the messages when session recover is called.
+     */
+    @Test(groups = "jmsListening", description = "Testing whether topic listening is working correctly without any "
+            + "exceptions in client ack mode")
+    public void topicListeningTestCase() {
+        try {
+            jmsTopicTransportListener.start(topicListeningParameters);
+            logger.info("JMS Transport Listener is starting to listen to the topic " + JMSTestConstants.TOPIC_NAME_1);
+            jmsServer.publishMessagesToTopic(JMSTestConstants.TOPIC_NAME_1);
+            Assert.assertTrue(topicTestMessageProcessor.getCount() > 10,
+                    "Expected message count is not received when " + "listening to topic "
+                            + JMSTestConstants.TOPIC_NAME_1);
+            jmsTopicTransportListener.stop();
+        } catch (Exception e) {
+            Assert.fail("Error while listening to topic");
```

Comment:
Simply throwing the error from the test case would fail it. We don't have to catch and Assert fail.